### PR TITLE
rand: instantiate the DRBGs upon first use.

### LINKED ISSUE #12714: RAND_status() changed behaviour after 3.0.0-alpha6
Perl's Net::SSLeay module runs [a test](https://github.com/radiator-software/p5-net-ssleay/blob/master/t/local/10_rand.t) that calls RAND_status() as soon as it starts. This returns successfully with old versions, such as OpenSSL 0.9.8f, up to 3.0.0-alpha6. With the current master RAND_status() returns 0. However, if RAND_bytes() is called first, it succeeds and RAND_status() the returns success. 

I did not tag this issue as a bug because I'm not completely sure what the expected behaviour is when RAND_status() is called before anything else. Some resources, such as https://wiki.openssl.org/index.php/Library_Initialization indicate that library initialises automatically with 1.1.0 and later. To skip any extra Perl and Net::SSLeay may introduce, I made a small tester, `rand_status.c`, which gives similar results on macOS Catalina and Ubuntu 18.04.

```
#include <stdio.h>
#include <openssl/rand.h>
#include <openssl/crypto.h>

int main(int argc, char **argv)
{
    int ret;
    unsigned char buf[10];

    if (argc > 1)
    {
        ret = RAND_bytes(buf, sizeof buf);
        printf("RAND_bytes : %d\n", ret);
    }

    ret = RAND_status();
    printf("RAND_status: %d\n", ret);

    printf("%s (Library: %s)\n", OPENSSL_VERSION_TEXT, SSLeay_version(SSLEAY_VERSION));
    return 0;
}
```

The tester simply calls RAND_status() and prints the return value. If called with any command line arguments, it calls RAND_bytes() first. The result of calling RAND_bytes() first is shown below with a test done against a build from today's master branch. RAND_status [man page](https://www.openssl.org/docs/manmaster/man3/RAND_status.html) indicates that RAND_bytes() should fail if RAND_status() does not return success, but seems like with the latest master this does not happen. A variation where nothing but RAND_status() is called twice, shows that both calls return 0.

Here are results from macOS. I hope this report is useful.

```
% cc -v
Apple clang version 11.0.3 (clang-1103.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
% uptime
18:25  up 28 days,  6:13, 33 users, load averages: 1.22 1.69 1.97
% rm rand_status

% cc -Wall -I $HOME/opt/openssl-0.9.8f/include -L $HOME/opt/openssl-0.9.8f/lib -o rand_status rand_status.c -lcrypto
% ./rand_status
RAND_status: 1
OpenSSL 0.9.8f 11 Oct 2007 (Library: OpenSSL 0.9.8f 11 Oct 2007)

% cc -Wall -I $HOME/opt/openssl-1.1.1g/include -L $HOME/opt/openssl-1.1.1g/lib -o rand_status rand_status.c -lcrypto
% ./rand_status                                                                                                     
RAND_status: 1
OpenSSL 1.1.1g  21 Apr 2020 (Library: OpenSSL 1.1.1g  21 Apr 2020)

% cc -Wall -I $HOME/opt/openssl-3.0.0-alpha6/include -L $HOME/opt/openssl-3.0.0-alpha6/lib -o rand_status rand_status.c -lcrypto 
% ./rand_status
RAND_status: 1
OpenSSL 3.0.0-alpha6 6 Aug 2020 (Library: OpenSSL 3.0.0-alpha6 6 Aug 2020)

% cc -Wall -I $HOME/opt/openssl-master/include -L $HOME/opt/openssl-master/lib -o rand_status rand_status.c -lcrypto 
% ./rand_status
RAND_status: 0
OpenSSL 3.0.0-alpha7-dev  (Library: OpenSSL 3.0.0-alpha7-dev )
% ./rand_status rand-bytes-first 
RAND_bytes : 1
RAND_status: 1
OpenSSL 3.0.0-alpha7-dev  (Library: OpenSSL 3.0.0-alpha7-dev )
```

### PR DESCRIPTION
**Pull request by paulidale**
For backwards compatibility, calling `RAND_status(3)` before generating any bytes needs to instantiate the primary DRBG.  This change instantiates the three DRBGs earlier, not relying on the automatic restart mechanism.

- [x] tests are added or updated

Fixes #12714

### CODE REVIEW

File: crypto/rand/rand_lib.c
Comment (mspncp) [inline crypto/rand/rand_lib.c:257]: Hmmm... actually my understanding was that RAND_status() should _not_ attempt an instantiate of an uninstantiated random generator, but only check the state. Instead, an automatic instantiation should happen on first use of `RAND_get0_primary()`, as it used to happen in `RAND_DRBG_get0_master()`, see my https://github.com/openssl/openssl/issues/12714#issuecomment-680283913:

> > RAND_status() must trigger implicit initialization the same way as RAND_bytes() does.
> 
> I don't recall that RAND_status() ever triggered an automatic instantiation and I don't think it should do so. The function is only meant to query the RNG's state. What actually is broken is the automatic instantiation of the primary (formerly: master) DRBG on startup. In prerelease alpha6, it is triggered on the first call to `RAND_DRBG_get0_master()` and carried out by `drbg_setup()`:

Comment (mspncp) [reply crypto/rand/rand_lib.c:257]: Just imagine that someone would call `EVP_RAND_state(RAND_get0_primary(NULL))` directly instead of RAND_status() and find out that the state is EVP_RAND_STATE_UNINITIALISED. That would be quite surprising, wouldn't it?

Comment (paulidale) [reply crypto/rand/rand_lib.c]: I misunderstood the intent.  Fix up pushed.

**Comment by paulidale** [comment]
Travis isn't relevant.

**Review by mspncp — APPROVED** [review]
Yes, now it's in the correct place to trigger initialization of the DRBG on first call (see call stack at the end).

Would you mind replacing "upon a status call" by "upon first use" (or something similar) in your commit message, because the change is not specific to the RAND_status() call?

Please also add a 'Fixes' tag.


```
rand: instantiate the primary DRBG upon first use

Fixes #12714
```


Approved, provided the commit message gets adjusted.



FWIW: here's the call stack of the instantiation on first use:

```
#0  prov_pool_acquire_entropy (pool=0x7b1000002540) at providers/implementations/rands/seeding/rand_unix.c:638
#1  0x00007ffff6fb9211 in prov_drbg_get_entropy (drbg=0x7b4400000140, pout=0x7fffffffdc78, entropy=256, min_len=32, max_len=2147483647, prediction_resistance=0) at providers/implementations/rands/drbg.c:210
#2  0x00007ffff6fb934e in get_entropy (drbg=0x7b4400000140, pout=0x7fffffffdc78, entropy=256, min_len=32, max_len=2147483647, prediction_resistance=0) at providers/implementations/rands/drbg.c:246
#3  0x00007ffff6fba110 in PROV_DRBG_instantiate (drbg=0x7b4400000140, strength=0, prediction_resistance=0, pers=0x7ffff7013f40 <ossl_pers_string> "OpenSSL NIST SP 800-90A DRBG", perslen=29) at providers/implementations/rands/drbg.c:477
#4  0x00007ffff6fab673 in drbg_ctr_instantiate_wrapper (vdrbg=0x7b4400000140, strength=0, prediction_resistance=0, pstr=0x0, pstr_len=0) at providers/implementations/rands/drbg_ctr.c:333
#5  0x00007ffff6dbde4c in evp_rand_instantiate_locked (ctx=0x7b04000018b0, strength=0, prediction_resistance=0, pstr=0x0, pstr_len=0) at crypto/evp/evp_rand.c:443
#6  0x00007ffff6dbdeb7 in EVP_RAND_instantiate (ctx=0x7b04000018b0, strength=0, prediction_resistance=0, pstr=0x0, pstr_len=0) at crypto/evp/evp_rand.c:455
#7  0x00007ffff6e5a4de in rand_new_drbg (libctx=0x0, parent=0x0, reseed_interval=256, reseed_time_interval=3600) at crypto/rand/rand_lib.c:472
#8  0x00007ffff6e5a5fd in RAND_get0_primary (ctx=0x0) at crypto/rand/rand_lib.c:496
#9  0x00007ffff6e59c07 in RAND_status () at crypto/rand/rand_lib.c:248
#10 0x00005555555551e6 in main (argc=1, argv=0x7fffffffe098) at rand_status.c:16
```

**Comment by paulidale** [comment]
I've updated the commit message.  I also removed _primary_ because all three get instantiated.

**Review by mspncp — APPROVED** [review]
Reconfirmed. Please remember to add the `Fixes #12714` line when merging.

**Comment by paulidale** [comment]
Isn't having the fixes line in the description here sufficient?  I didn't think it needed to be in the commit message.

**Comment by mspncp** [comment]
> Isn't having the fixes line in the description here sufficient? I didn't think it needed to be in the commit message.

The former might be sufficient for GitHub's autoclose feature, but a human trying to understand a code change or  finding a bug might prefer to use `git log`, `git annotate`, and `git bisect` in his local working copy. So it's in the git log close to the code where the relevant information should be found and not on GitHub IMHO. But YMMV of course.

**Comment by mspncp** [comment]
P.S: I often search the git log for issue and pr numbers

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Added the reporting PR number to the git commit message.  I don't consider this enough of a change to require approval again.

**Comment by paulidale** [comment]
Merged to master thanks.

**Pull request closed without merging**
